src/net/zephyr/network.c

```diff
@@ -0,0 +1,389 @@
+        return r;
+    }
+*/
+
+    memset(&serv_addr, 0, sizeof(serv_addr));
+    serv_addr.sin_family = AF_INET;
+    serv_addr.sin_port = htons(port);
+
+    if (inet_pton(AF_INET, ip_addr, &serv_addr.sin_addr) <= 0)
+    {
+        r.tag = _z_res_t_ERR;
+        r.value.error = _zn_err_t_INVALID_LOCATOR;
+        r.value.socket = 0;
+        return r;
+    }
+
+    if (connect(r.value.socket, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0)
+    {
+        r.tag = _z_res_t_ERR;
+        r.value.error = _zn_err_t_TX_CONNECTION;
+        r.value.socket = 0;
+        return r;
+    }
+
+    return r;
+}
+
+void _zn_close_tx_session(_zn_socket_t sock)
+{
+    shutdown(sock, 2);
+}
+
+/*------------------ Datagram ------------------*/
+int _zn_send_dgram_to(_zn_socket_t sock, const _z_wbuf_t *wbf, const struct sockaddr *dest, socklen_t salen)
+{
+    _Z_DEBUG("Sending data on socket....\n");
+    int wb = 0;
+    for (size_t i = 0; i < _z_wbuf_len_iosli(wbf); i++)
+    {
+        z_bytes_t a = _z_iosli_to_bytes(_z_wbuf_get_iosli(wbf, i));
+        int res = sendto(sock, a.val, a.len, 0, dest, salen);
+        _Z_DEBUG_VA("Socket returned: %d\n", wb);
+        if (res <= 0)
+        {
+            _Z_DEBUG_VA("Error while sending data over socket [%d]\n", wb);
+            return -1;
+        }
+        wb += res;
+    }
+    return wb;
+}
+
+int _zn_recv_dgram_from(_zn_socket_t sock, _z_rbuf_t *rbf, struct sockaddr *from, socklen_t *salen)
+{
+    int rb = recvfrom(sock, _z_rbuf_get_wptr(rbf), _z_rbuf_space_left(rbf), 0, from, salen);
+    if (rb > 0)
+        _z_rbuf_set_wpos(rbf, _z_rbuf_get_wpos(rbf) + rb);
+
+    return rb;
+}
+
+/*------------------ Receive ------------------*/
+int _zn_recv_bytes(_zn_socket_t sock, uint8_t *ptr, size_t len)
+{
+    int n = len;
+    int rb;
+
+    do
+    {
+        rb = recv(sock, ptr, n, 0);
+        if (rb == 0)
+            return -1;
+        n -= rb;
+        ptr = ptr + (len - n);
+    } while (n > 0);
+
+    return 0;
+}
+
+int _zn_recv_rbuf(_zn_socket_t sock, _z_rbuf_t *rbf)
+{
+    int rb = recv(sock, _z_rbuf_get_wptr(rbf), _z_rbuf_space_left(rbf), 0);
+    if (rb > 0)
+        _z_rbuf_set_wpos(rbf, _z_rbuf_get_wpos(rbf) + rb);
+    return rb;
+}
+
+/*------------------ Send ------------------*/
+int _zn_send_wbuf(_zn_socket_t sock, const _z_wbuf_t *wbf)
+{
+    for (size_t i = 0; i < _z_wbuf_len_iosli(wbf); i++)
+    {
+        z_bytes_t bs = _z_iosli_to_bytes(_z_wbuf_get_iosli(wbf, i));
+        int n = bs.len;
+        int wb;
+        do
+        {
+            _Z_DEBUG("Sending wbuf on socket...");
+            wb = send(sock, bs.val, n, 0);
+            _Z_DEBUG_VA(" sent %d bytes\n", wb);
+            if (wb <= 0)
+            {
+                _Z_DEBUG_VA("Error while sending data over socket [%d]\n", wb);
+                return -1;
+            }
+            n -= wb;
+            bs.val += bs.len - n;
+        } while (n > 0);
+    }
+
+    return 0;
+}
+
```

Comment:
I would simply remove the commented code and leave the note saying that iovec is not supported in Zephyr.